src/lib.rs

```diff
@@ -82,101 +82,8 @@ pub trait WriteBase256 {
     fn write_u8(&mut self, data: u8) -> Result<(), Self::Error> { self.write(&[data]) }
 }
 
-const CHECKSUM_LENGTH: usize = 6;
-
-/// Allocationless Bech32 writer that accumulates the checksum data internally and writes them out
-/// in the end.
-pub struct Bech32Writer<'a, Ck: Checksum> {
-    formatter: &'a mut dyn fmt::Write,
-    engine: checksum::Engine<Ck>,
-}
-
-impl<'a, Ck: Checksum> Bech32Writer<'a, Ck> {
-    /// Creates a new writer that can write a bech32 string without allocating itself.
-    ///
-    /// This is a rather low-level API and doesn't check the HRP or data length for standard
-    /// compliance.
-    fn new(hrp: Hrp, fmt: &'a mut dyn fmt::Write) -> Result<Bech32Writer<'a, Ck>, fmt::Error> {
-        let mut engine = checksum::Engine::new();
-        engine.input_hrp(&hrp);
-
-        for c in hrp.lowercase_char_iter() {
-            fmt.write_char(c)?;
-        }
-        fmt.write_char(SEP)?;
-
-        Ok(Bech32Writer { formatter: fmt, engine })
-    }
-
-    /// Writes out the checksum at the end.
-    ///
-    /// If this method isn't explicitly called this will happen on drop.
-    pub fn finalize(mut self) -> fmt::Result {
-        self.write_checksum()?;
-        mem::forget(self);
-        Ok(())
-    }
-
-    /// Calculates and writes a checksum to `self`.
-    fn write_checksum(&mut self) -> fmt::Result {
-        self.engine.input_target_residue();
-
-        let mut checksum_remaining = self::CHECKSUM_LENGTH;
-        while checksum_remaining > 0 {
-            checksum_remaining -= 1;
-
-            let fe = u5::try_from(self.engine.residue().unpack(checksum_remaining))
-                .expect("unpack returns valid field element");
-            self.formatter.write_char(fe.to_char())?;
-        }
-
-        Ok(())
-    }
-}
-
-impl<'a, Ck: Checksum> WriteBase32 for Bech32Writer<'a, Ck> {
-    type Error = fmt::Error;
-
-    fn write_u5(&mut self, data: u5) -> fmt::Result {
-        self.engine.input_fe(data);
-        self.formatter.write_char(data.to_char())
-    }
-}
-
-impl<'a, Ck: Checksum> Drop for Bech32Writer<'a, Ck> {
-    fn drop(&mut self) {
-        self.write_checksum().expect("Unhandled error writing the checksum on drop.")
-    }
-}
-
-/// Parses/converts base32 slice to `Self`.
-///
-/// This trait is the reciprocal of `ToBase32`.
-pub trait FromBase32: Sized {
-    /// The associated error which can be returned from parsing (e.g. because of bad padding).
-    type Error;
-
-    /// Converts a base32 slice to `Self`.
-    fn from_base32(b32: &[u5]) -> Result<Self, Self::Error>;
-}
-
 macro_rules! write_base_n {
```

Comment:
In e2527dc19c3e8e9a4cd678a955fc5b161e175c87:

Might as well delete this macro, and the `WriteBase*` traits, so we can have a solid sea of red here. You've removed every usage of these traits and arleady crossed the "breaking API change" rubicon.

Reply:
This requires you delete the `convert_bits` function and all its unit tests. I'll leave it up to you if you want to do that here. It's not necessary for this PR to compile so maybe you want to keep the diff minimal.

Ditto the `CheckBase32` trait and its tests.